hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala

```diff
@@ -367,4 +375,82 @@ final case class PCanonicalArray(elementType: PType, required: Boolean = false)
       destOff
     )
   }
+
+  override def storeShallowAtOffset(dstAddress: Code[Long], valueAddress: Code[Long]): Code[Unit] =
+    Region.storeAddress(dstAddress, valueAddress)
+
+  override def storeShallowAtOffset(dstAddress: Long, valueAddress: Long) {
+    Region.storeAddress(dstAddress, valueAddress)
+  }
+
+  // semantically this function expects a non-null sourceAddress, and by that property, this function results in a non-null value
+  override def copyFromType(mb: MethodBuilder, region: Code[Region], srcPType: PType, srcAddress: Code[Long],
+  allowDowncast: Boolean = false, forceDeep: Boolean = false): Code[Long] = {
+    assert(srcPType.isInstanceOf[PArray])
+
+    val sourceType = srcPType.asInstanceOf[PArray]
+    val sourceElementType = sourceType.elementType.fundamentalType
+    val destElementType = this.elementType.fundamentalType
+
+    if (sourceElementType != destElementType) {
+      assert(sourceElementType isOfType destElementType)
```

Comment:
I find it somewhat hard to follow the control flow here, in part due to checking negatives. I think the following would be clearer:

```scala
    if (sourceElementType == destElementType) {
      if (forceDeep)
        if(sourceElementType.isPrimitive)
          return this.copyFrom(mb, region, srcAddress)
        else
          return <copy and fixup>
      return srcAddress
    }
```